Bad SPARQLStore handling of empty literals

When an empty literal is returned to a SPARQLStore, it becomes an `rdflib.term.Literal(u'None')` instead of an `rdflib.term.Literal(u'')`:

```python
# ./fuseki-server --update --mem /empty_literal

import rdflib
from rdflib.plugins.stores.sparqlstore import SPARQLUpdateStore

store = SPARQLUpdateStore('http://fuseki.til.io:3030/empty_literal/query',
                          'http://fuseki.til.io:3030/empty_literal/update')

graph = rdflib.Graph(store, 'http://example.com/')

graph.add((rdflib.URIRef('http://example.com/s'), 
           rdflib.URIRef('http://example.com/p'), 
           rdflib.Literal('')))

o = tuple(graph)[0][2]
assert o == rdflib.Literal(''), repr(o)
```
raises an `AssertionError: rdflib.term.Literal(u'None')`



The problem seems to be in `rdflib.plugins.stores.sparqlstore:CastToTerm`:

```python
import io
from xml.etree import ElementTree

import rdflib
from rdflib.plugins.stores.sparqlstore import CastToTerm

s = """<ns0:literal xmlns:ns0="http://www.w3.org/2005/sparql-results#" />"""
node = ElementTree.parse(io.BytesIO(s)).getroot()
term = CastToTerm(node)
assert term  == rdflib.Literal(''), repr(term)
```
raises an `AssertionError: rdflib.term.Literal(u'None')`

--- a/rdflib/plugins/stores/sparqlstore.py
+++ b/rdflib/plugins/stores/sparqlstore.py
@@ -132,10 +132,7 @@ def CastToTerm(node):
         value = node.text if node.text is not None else ''
         if 'datatype' in node.attrib:
             dT = URIRef(node.attrib['datatype'])
-            if False:  # not node.xpath('*'):
-                return Literal('', datatype=dT)
-            else:
-                return Literal(value, datatype=dT)
+            return Literal(value, datatype=dT)
         elif '{http://www.w3.org/XML/1998/namespace}lang' in node.attrib:
             return Literal(value, lang=node.attrib[
                 "{http://www.w3.org/XML/1998/namespace}lang"])

--- a/rdflib/plugins/stores/sparqlstore.py
+++ b/rdflib/plugins/stores/sparqlstore.py
@@ -129,17 +129,18 @@ def CastToTerm(node):
     elif node.tag == '{%s}uri' % SPARQL_NS:
         return URIRef(node.text)
     elif node.tag == '{%s}literal' % SPARQL_NS:
+        value = node.text if node.text is not None else ''
         if 'datatype' in node.attrib:
             dT = URIRef(node.attrib['datatype'])
             if False:  # not node.xpath('*'):
                 return Literal('', datatype=dT)
             else:
-                return Literal(node.text, datatype=dT)
+                return Literal(value, datatype=dT)
         elif '{http://www.w3.org/XML/1998/namespace}lang' in node.attrib:
-            return Literal(node.text, lang=node.attrib[
+            return Literal(value, lang=node.attrib[
                 "{http://www.w3.org/XML/1998/namespace}lang"])
         else:
-            return Literal(node.text)
+            return Literal(value)
     else:
         raise Exception('Unknown answer type')
 

--- a/test/test_issue457.py
+++ b/test/test_issue457.py
@@ -5,8 +5,9 @@ from xml.etree import ElementTree
 
 import rdflib
 from rdflib.plugins.stores.sparqlstore import CastToTerm
+from rdflib.py3compat import b
 
-s = """<ns0:literal xmlns:ns0="http://www.w3.org/2005/sparql-results#" />"""
+s = b("""<ns0:literal xmlns:ns0="http://www.w3.org/2005/sparql-results#" />""")
 node = ElementTree.parse(io.BytesIO(s)).getroot()
 term = CastToTerm(node)
 assert term  == rdflib.Literal(''), repr(term)

--- a/test/test_issue457.py
+++ b/test/test_issue457.py
@@ -1,0 +1,13 @@
+# test for https://github.com/RDFLib/rdflib/issues/457
+
+import io
+from xml.etree import ElementTree
+
+import rdflib
+from rdflib.plugins.stores.sparqlstore import CastToTerm
+
+s = """<ns0:literal xmlns:ns0="http://www.w3.org/2005/sparql-results#" />"""
+node = ElementTree.parse(io.BytesIO(s)).getroot()
+term = CastToTerm(node)
+assert term  == rdflib.Literal(''), repr(term)
+

--- a/test/test_sparqlupdatestore.py
+++ b/test/test_sparqlupdatestore.py
@@ -2,7 +2,7 @@
 import unittest
 import re
 
-from rdflib import ConjunctiveGraph, URIRef
+from rdflib import ConjunctiveGraph, URIRef, Literal
 from rdflib.util import from_n3
 
 # this assumed SPARQL1.1 query/update endpoints
@@ -105,23 +105,23 @@ class TestSparql11(unittest.TestCase):
 
     def testUpdate(self):
         self.graph.update("INSERT DATA { GRAPH <urn:graph> { <urn:michel> <urn:likes> <urn:pizza> . } }")
-        
+
         g = self.graph.get_context(graphuri)
         self.assertEquals(1, len(g), 'graph contains 1 triples')
-        
+
     def testUpdateWithInitNs(self):
         self.graph.update(
             "INSERT DATA { GRAPH ns:graph { ns:michel ns:likes ns:pizza . } }",
             initNs={'ns': URIRef('urn:')}
         )
-        
+
         g = self.graph.get_context(graphuri)
         self.assertEquals(
             set(g.triples((None,None,None))),
             set([(michel,likes,pizza)]),
             'only michel likes pizza'
         )
-        
+
     def testUpdateWithInitBindings(self):
         self.graph.update(
             "INSERT { GRAPH <urn:graph> { ?a ?b ?c . } } WherE { }",
@@ -131,7 +131,7 @@ class TestSparql11(unittest.TestCase):
                 'c': URIRef('urn:pizza'),
             }
         )
-        
+
         g = self.graph.get_context(graphuri)
         self.assertEquals(
             set(g.triples((None,None,None))),
@@ -150,7 +150,7 @@ class TestSparql11(unittest.TestCase):
                 'd': URIRef('urn:bob'),
             }
         )
-        
+
         g = self.graph.get_context(graphuri)
         self.assertEquals(
             set(g.triples((None,None,None))),
@@ -268,6 +268,18 @@ class TestSparql11(unittest.TestCase):
         if empty_graph_iri in named_graphs:
             self.assertTrue(empty_graph_iri in [unicode(g.identifier)
                                                 for g in self.graph.contexts()])
+
+    def testEmptyLiteral(self):
+        # test for https://github.com/RDFLib/rdflib/issues/457
+        # also see test_issue457.py which is sparql store independent!
+        g = self.graph.get_context(graphuri)
+        g.add((
+            URIRef('http://example.com/s'),
+            URIRef('http://example.com/p'),
+            Literal('')))
+
+        o = tuple(g)[0][2]
+        self.assertEquals(o, Literal(''), repr(o))
 
 from nose import SkipTest
 import urllib2

